Doc/library/sqlite3.rst

```diff
@@ -2614,6 +2596,65 @@ With some adjustments, the above recipe can be adapted to use a
 instead of a :class:`~collections.namedtuple`.
 
 
+.. _sqlite3-howto-encoding:
+
+How to handle non-UTF-8 text encodings
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for UTF-8 encoded text, but it will fail for other encodings
+and invalid UTF-8.
+You can use a custom :attr:`~Connection.text_factory` to handle such cases.
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing non-UTF-8 encodings,
+or even arbitrary data.
+To demonstrate, let's assume we build a database by importing a ISO-8859-2
+(Latin 2) encoded CSV file, for example a list of Czech-English dictionary
+entries, :file:`entries.csv`:
+
+.. code-block:: shell
+
+   $ python3
+   >>> with open("entries.csv", "w", encoding="latin2") as f:
+   ...     f.write("czech,english\n")  # header row
+   ...     f.write("položka,entry\n")  # data rows
+   ...     f.write("částka,sum\n")
+   >>> quit()
+   $ sqlite3 dictionary.db
+   sqlite> .mode csv
+   sqlite> .import entries.csv dict
+   sqlite> .quit
+
+:file:`dictionary.db` now contains a database that contains Latin 2 encoded text.
+Assuming we now have a :class:`Connection` instance :py:data:`!con`
+connected to :file:`dictionary.db`,
+we can decode the Latin 2 encoded text using this :attr:`~Connection.text_factory`:
+
+.. testcode::
+
+   con.text_factory = lambda data: str(data, encoding="latin2")
+
+For invalid UTF-8 or arbitrary data in stored in ``TEXT`` table columns,
+you can use the following technique, borrowed from the :ref:`unicode-howto`:
+
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+
+.. note::
+
+   Strings containing surrogate escapes and must be treated with care.
+   You cannot simply pass them back to SQLite,
+   for example using :meth:`~Cursor.execute`,
+   since the :mod:`!sqlite3` module APIs only accept UTF-8 encoded strings.
```

Comment:
@serhiy-storchaka

Alt 1:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite.
```

Alt 2:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite;
   the :mod:`!sqlite3` module API does not understand
   surrogate escape codes.
```

Other suggestions?

Reply:
Alt 3: Just remove the note.

Reply:
My 2 cents
Alt 4:
```suggestion
   Strings containing surrogate escapes must be treated with care as
   modules that attempt to encode them require support for surrogate
   escapes to handle them correctly.
   The :mod:`!sqlite3` module API does not support surrogate escapes
   so these strings cannot be passed back to the :mod:`!sqlite3` module API.
```

Reply:
```suggestion
   Strings containing surrogate escapes and must be treated with care.
   For example, you cannot simply pass them back to SQLite,
   since the :mod:`!sqlite3` module API does not support strings
   containing surrogate escape codes.
```